File: packages/shared-python/townops_shared/utils/rabbitmq.py
import os
from collections.abc import Awaitable, Callable

import aio_pika
import structlog

log = structlog.get_logger(__name__)


class RabbitMQClient:
  def __init__(self) -> None:
    self.connection: aio_pika.abc.AbstractConnection | None = None
    self.channel: aio_pika.abc.AbstractChannel | None = None

  async def connect(self) -> None:
    """
    Establishes a robust connection to RabbitMQ to survive network drops.
    """
    if self.connection and not self.connection.is_closed:
      return

    url = os.getenv("RABBITMQ_URL")
    if not url:
      raise ValueError("RABBITMQ_URL is not set")  # noqa: TRY003

    self.connection = await aio_pika.connect_robust(
      url, client_properties={"connection_name": "townops-composite"}
    )
    self.channel = await self.connection.channel()
    log.info("RabbitMQ connection established")

  async def disconnect(self) -> None:
    """
    Closes the RabbitMQ connection.
    """
    if self.connection:
      await self.connection.close()
      log.info("RabbitMQ connection closed")

  async def declare_exchange(
    self, name: str, exchange_type: str = "topic"
  ) -> aio_pika.abc.AbstractExchange:
    """
    Declares an exchange for publishing messages.
    """
    if not self.channel:
      await self.connect()
    if self.channel is None:
      raise RuntimeError("RabbitMQ channel not connected")  # noqa: TRY003
    return await self.channel.declare_exchange(name, exchange_type, durable=True)

  async def publish(
    self,
    exchange_name: str,
    routing_key: str,
    message_body: bytes,
    content_type: str = "application/json",
  ) -> None:
    """
    Publishes a message to an exchange.
    """
    if not self.channel:
      await self.connect()
    if self.channel is None:
      raise RuntimeError("RabbitMQ channel not connected")  # noqa: TRY003

    exchange = await self.channel.get_exchange(exchange_name)
    await exchange.publish(
      aio_pika.Message(
        body=message_body,
        content_type=content_type,
      ),
      routing_key=routing_key,
    )
    log.debug("Message published", exchange=exchange_name, routing_key=routing_key)

  async def consume(
    self,
    queue_name: str,
    callback: Callable[[aio_pika.abc.AbstractIncomingMessage], Awaitable[None]],
    exchange_name: str | None = None,
    routing_key: str | None = None,
    arguments: dict[str, any] | None = None,
  ) -> None:
    """
    Continuously consumes from a queue and executes an async callback for each message.
    Optional binding to an exchange with optional arguments for queue declaration.
    """
    if not self.channel:
      await self.connect()
    if self.channel is None:
      raise RuntimeError("RabbitMQ channel not connected")  # noqa: TRY003

    await self.channel.set_qos(prefetch_count=1)
    queue = await self.channel.declare_queue(
      queue_name, durable=True, arguments=arguments
    )

    if exchange_name and routing_key:
      exchange = await self.channel.declare_exchange(
        exchange_name, "topic", durable=True
      )
      await queue.bind(exchange, routing_key=routing_key)
      log.info(
        "Queue bound to exchange",
        queue=queue_name,
        exchange=exchange_name,
        routing_key=routing_key,
      )

    log.info("Starting consumer", queue=queue_name)
    async with queue.iterator() as queue_iter:
      async for message in queue_iter:
        # message.process() automatically ACKs on exit if no exception was raised,
        # or NACKs with requeue if an exception is raised inside the context block.
        async with message.process():
          try:
            await callback(message)
          except Exception as e:
            log.exception(
              "Error processing message",
              error=str(e),
              queue=queue_name,
              message_id=message.message_id,
            )
            # Re-raise to trigger NACK and potential requeue
            raise

File: packages/shared-ts/src/rabbitmq.ts
import { AMQPClient } from "@cloudamqp/amqp-client";
import type { AMQPChannel, AMQPMessage } from "@cloudamqp/amqp-client";

import { env } from "./env";
import { logger } from "./logger";

type QueueName =
  | "assign-job-queue"
  | "alert-queue"
  | "metrics-queue"
  | "handle-breach-queue"
  | "error-audit-queue"
  | "sla-timers-queue";

const DEFAULT_QUEUE_ARGUMENTS: Record<
  QueueName,
  Record<string, string> | undefined
> = {
  "assign-job-queue": { "x-dead-letter-exchange": "townops.dlx" },
  "alert-queue": { "x-dead-letter-exchange": "townops.dlx" },
  "metrics-queue": { "x-dead-letter-exchange": "townops.dlx" },
  "handle-breach-queue": { "x-dead-letter-exchange": "townops.dlx" },
  "error-audit-queue": undefined,
  "sla-timers-queue": {
    "x-dead-letter-exchange": "townops.events",
    "x-dead-letter-routing-key": "sla.breached",
  },
};

/**
 * @class RabbitMQClient
 * @description A client for interacting with RabbitMQ.
 */
class RabbitMQClient {
  private client: AMQPClient | null = null;
  private connection: Awaited<ReturnType<AMQPClient["connect"]>> | null = null;
  private channel: AMQPChannel | null = null;

  /**
   * @method connect
   * @description Establishes a connection to RabbitMQ.
   */
  async connect(): Promise<void> {
    if (this.channel) return;

    const url = env.RABBITMQ_URL;
    if (!url) {
      throw new Error("RABBITMQ_URL environment variable is not set");
    }

    try {
      this.client = new AMQPClient(url);
      this.connection = await this.client.connect();
      this.channel = await this.connection.channel();
      logger.info("RabbitMQ connection established");
    } catch (e) {
      logger.error({ err: e }, "Failed to connect to RabbitMQ");
      throw e;
    }
  }

  /**
   * @method disconnect
   * @description Closes the RabbitMQ connection.
   */
  async disconnect(): Promise<void> {
    if (this.connection) {
      await this.connection.close();
      this.channel = null;
      this.connection = null;
      logger.info("RabbitMQ connection closed");
    }
  }

  /**
   * @method declareExchange
   * @description Declares an exchange if it does not exist.
   */
  async declareExchange(
    name: string,
    type: "direct" | "fanout" | "topic" | "headers" = "topic"
  ): Promise<void> {
    if (!this.channel) await this.connect();
    const channel = this.channel;
    if (!channel) throw new Error("RabbitMQ channel not connected");
    await channel.exchangeDeclare(name, type, { durable: true });
  }

  /**
   * @method publish
   * @param exchangeName - The name of the exchange to publish to.
   * @param routingKey - The routing key to publish to.
   * @param message - The message to publish.
   * @param options - Optional parameters for publishing the message.
   * @description Publishes a message to an exchange.
   */
  async publish(
    exchangeName: "townops.events" | "townops.dlx",
    routingKey:
      | "case.opened"
      | "case.escalated"
      | "case.no_access"
      | "job.assigned"
      | "job.done"
      | "sla.breached"
      | "#",
    message: string | object,
    options?: { contentType?: string }
  ): Promise<void> {
    if (!this.channel) await this.connect();
    const channel = this.channel;
    if (!channel) throw new Error("RabbitMQ channel not connected");

    const data =
      typeof message === "string" ? message : JSON.stringify(message);
    const contentType =
      options?.contentType ||
      (typeof message === "object" ? "application/json" : "text/plain");

    await channel.basicPublish(exchangeName, routingKey, data, {
      contentType,
      deliveryMode: 2, // 2 = persistent - survives broker restarts
    });
    logger.debug({ exchangeName, routingKey }, "Message published to RabbitMQ");
  }

  /**
   * @method consume
   * @param queueName - The name of the queue to consume from.
   * @param callback - The callback function to handle messages.
   * @param options - Optional parameters for consuming messages.
   * @description Consumes messages from a queue with callback handler and automatic/requeue ACK handling.
   */
  async consume(
    queueName: QueueName,
    callback: (message: AMQPMessage) => Promise<void>,
    options?: {
      exchangeName?: "townops.events" | "townops.dlx";
      routingKey?:
        | "case.opened"
        | "case.escalated"
        | "case.no_access"
        | "job.assigned"
        | "job.done"
        | "sla.breached"
        | "#";
      arguments?: Record<string, any>;
    }
  ): Promise<unknown> {
    if (!this.channel) await this.connect();
    const channel = this.channel;
    if (!channel) throw new Error("RabbitMQ channel not connected");

    // Set prefetch to 1 for fair dispatching
    await channel.basicQos(1);

    const queueArguments = DEFAULT_QUEUE_ARGUMENTS[queueName]
      ? { ...DEFAULT_QUEUE_ARGUMENTS[queueName], ...options?.arguments }
      : options?.arguments;

    const q = await channel.queue(queueName, { durable: true }, queueArguments);

    // Optional binding to an exchange
    if (options?.exchangeName && options.routingKey) {
      await channel.exchangeDeclare(options.exchangeName, "topic", {
        durable: true,
      });
      await channel.queueBind(
        queueName,
        options.exchangeName,
        options.routingKey
      );
      logger.info(
        {
          queueName,
          exchangeName: options.exchangeName,
          routingKey: options.routingKey,
        },
        "Bound queue to exchange"
      );
    }

    logger.info({ queueName }, "Starting RabbitMQ consumer");

    const consumer = await q.subscribe(
      { noAck: false },
      async (msg: AMQPMessage) => {
        try {
          await callback(msg);
          await msg.ack();
        } catch (e) {
          logger.error(
            { err: e, queueName },
            "Error processing RabbitMQ message"
          );
          // Requeue set to false, so it gets handled by DLX if configured
          await msg.nack(true, false);
        }
      }
    );

    return consumer;
  }
}

/**
 * @description A singleton instance of RabbitMQClient.
 */
export const rabbitmqClient = new RabbitMQClient();
